lib/pacemaker/pcmk_sched_group.c

```diff
@@ -350,87 +350,131 @@ pcmk__group_internal_constraints(pe_resource_t *rsc)
 
     if (pe__group_flag_is_set(dependent, pe__group_colocated)) {
+        // Colocate first member (internal colocations will handle the rest)
         member = (pe_resource_t *) dependent->children->data;
         member->cmds->apply_coloc_score(member, primary, colocation, true);
         return;
+    }
 
-    } else if (colocation->score >= INFINITY) {
-        pcmk__config_err("%s: Cannot perform mandatory colocation "
-                         "between non-colocated group and %s",
+    if (colocation->score >= INFINITY) {
+        pcmk__config_err("%s: Cannot perform mandatory colocation between "
+                         "non-colocated group and %s",
                          dependent->id, primary->id);
         return;
     }
 
-    for (; gIter != NULL; gIter = gIter->next) {
-        pe_resource_t *child_rsc = (pe_resource_t *) gIter->data;
-
-        child_rsc->cmds->apply_coloc_score(child_rsc, primary, colocation,
-                                           true);
+    // Colocate each member individually
+    for (GList *iter = dependent->children; iter != NULL; iter = iter->next) {
+        member = (pe_resource_t *) iter->data;
+        member->cmds->apply_coloc_score(member, primary, colocation, true);
     }
-    return;
+}
 
-for_primary:
-    gIter = primary->children;
-    CRM_CHECK(dependent->variant == pe_native, return);
+/*!
+ * \internal
+ * \brief Apply a colocation's score to node weights or resource priority
+ *
+ * Given a colocation constraint for some other resource with a group, apply the
+ * score to the dependent's allowed node weights (if we are still placing
+ * resources) or priority (if we are choosing promotable clone instance roles).
+ *
+ * \param[in,out] dependent      Dependent resource in colocation
+ * \param[in]     primary        Primary group resource in colocation
+ * \param[in]     colocation     Colocation constraint to apply
+ */
+static void
+colocate_with_group(pe_resource_t *dependent, const pe_resource_t *primary,
+                    const pcmk__colocation_t *colocation)
+{
+    pe_resource_t *member = NULL;
 
     pe_rsc_trace(primary,
                  "Processing colocation %s (%s with group %s) for primary",
                  colocation->id, dependent->id, primary->id);
 
-    member = (pe_resource_t *) primary->children->data;
-
     if (pcmk_is_set(primary->flags, pe_rsc_provisional)) {
         return;
+    }
+
+    if (pe__group_flag_is_set(primary, pe__group_colocated)) {
 
-    } else if (pe__group_flag_is_set(primary, pe__group_colocated)
-               && (member != NULL)) {
         if (colocation->score >= INFINITY) {
             // Dependent can't start until group is fully up
             member = pe__last_group_member(primary);
-        } // else dependent can start as long as group is partially up
+        } else {
+            // Dependent can start as long as group is partially up
+            member = (pe_resource_t *) primary->children->data;
+        }
```

Comment:
Looking at it again, the comments are what's questionable ... ordering is not involved here.

The difference is whether the first member or last member is used as the primary in apply_coloc_score(). Only primitives can be group members, so it's using pcmk__primitive_apply_coloc_score() either way. The *colocation's* score is added to the dependent's score on the group member's node.

It looks like the only significant difference would be when the first member is assignable to a node (and in specified role if any) but a later member is not. For optional colocations, the colocation score would be applied, whereas for mandatory colocations, the dependent couldn't be placed at all.

That seems reasonable. The comment should be clarified. I think the documentation is fine since in the end the effect really is just optional vs. mandatory.

Reply:
Added commit on top to update comment